streamingest: periodically output replication-frontier.txt file

adityamaru opened the pull request:
Please see individual commits for details.

Informs: #108374

adityamaru committed:
streamingccl: persist partition specs in job_info

This change adds the source and destination sql instance IDs
to the StreamIngestionPartitionSpec. It also teaches the
planning logic to persist the partition specs in the job_info
table so that it can be accessed by the frontier processor in
follow up commits.

Informs: #108374
Release note: None

adityamaru committed:
streamingest: periodically persist the frontier to the job_info table

This change teaches the stream ingestion frontier to periodically
persist a snapshot of the frontier to the job_info table. This snapshot
is persisted at a fixed info key and so as per the job_info table
semantics a new write will delete and insert a new row at the info key.
This ensures that we only store the latest view of the frontier at a given
point in time.

This frontier in addition to the StreamIngestionPartitionSpecs that
are peristed in the first commit can now be used by the ingestion resumer
to construct a mapping from:
src ID -> dest ID -> frontier ts -> span

This will be done in a follow up commit.

Informs: #108374
Release note: None

adityamaru committed:
streamingest: output `replication-frontier.txt` file on demand

This is the last commit in the series that uses the two pieces
of information persisted in the previous commits. Namely,

- The StreamIngestionPartitionSpec of each partition providing a mapping from
span to src and dest SQLInstanceID.

- The snapshot of the frontier tracking how far each span has been replicated
up to.

The file is generated everytime the destination job is requested for its execution
details. The file will be consumable from the `Advanced Debugging` tab on the
job details page and will be of the form:

```
Src Instance Dest Instance Span                    Frontier Timestamp                   Behind By
1            1             /Tenant/2{-/Table/106}  2023-08-16 19:22:44.940537 +0000 UTC 13s
1            1             /Tenant/{2/Table/106-3} 2023-08-16 19:22:44.739163 +0000 UTC 13.2s
```

Informs: #108374
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/108863)
<!-- Reviewable:end -->

dt commented on pkg/ccl/streamingccl/streamingest/stream_ingestion_dist.go:
```diff
@@ -55,6 +62,18 @@ var replanFrequency = settings.RegisterDurationSetting(
 	settings.PositiveDuration,
 )
 
+var persistExecutionDetailsFrequency = settings.RegisterDurationSetting(
```
I'd prefer not to add a cluster setting for this, but rather just leave it as a const or env var for now or make it an option on the job.

If we must use a cluster setting, it would only be for the system tenant, right?

adityamaru replied:
changed it to a const = 10 * time.Minute for now. This will go away once we have the resumer specific method.

msbutler commented on pkg/ccl/streamingccl/streamingest/stream_ingestion_dist.go:
what do you mean by "timer-based loop"? This looks like a timer based loop to me?

adityamaru replied:
Yeah, the TODO is to _replace_ the time-based loop with a fetch model once the requesting of execution details once each resumer can be contacted to be provide its relevant details.

msbutler replied:
lol, I totally missed the "replace" in that docstring.

msbutler commented on pkg/ccl/streamingccl/streamingest/stream_ingestion_dist.go:
nit: grab the cluster setting within the loop so people can update the setting without pausing the job

msbutler replied:
or if you want to be fancy, you could use the `SetOnChange` api. E[xample](https://github.com/cockroachdb/cockroach/blob/ef850fb9773921d19cfa220fd756623c4d91505c/pkg/spanconfig/spanconfigmanager/manager.go#L120)

adityamaru replied:
changed it to a const as dt recommended above, this is going away very soon once the remaining pieces of https://github.com/cockroachdb/cockroach/issues/109671 are implemented.

msbutler commented on pkg/ccl/streamingccl/streamingest/stream_ingestion_dist.go:
```diff
@@ -191,13 +236,134 @@ func (p *replicationFlowPlanner) containsInitialStreamAddresses() bool {
 	return len(p.initialStreamAddresses) > 0
 }
 
+type frontierExecutionDetails struct {
```
nit: put all of these helper funcs in a new go file?

msbutler commented on pkg/ccl/streamingccl/streamingest/stream_ingestion_dist_test.go:
```diff
@@ -316,3 +316,139 @@ func TestSourceDestMatching(t *testing.T) {
+						SrcInstanceID:  1,
+						DestInstanceID: 2,
+						Spans: []roachpb.Span{
+							{Key: roachpb.Key("b"), EndKey: roachpb.Key("d")},
```
nit: i don't think this is a realistic test case: two different partition specs will never have the same src-dest pair. 

I could see two useful related test cases:
1.  Remove one of the partition specs, and add to the `Spans` list?
2. keep what you hav, and change one of the DestIDs to 3

msbutler replied:
sounds good!

msbutler commented on pkg/ccl/streamingccl/streamingest/stream_ingestion_dist_test.go:
```diff
@@ -316,3 +316,139 @@ func TestSourceDestMatching(t *testing.T) {
+					destInstanceID: 2,
+					span:           "{b-d}",
+				}},
+		},
```
could you add a test case where a partitionSpec's Span has a split frontier? i.e.
```
partitionSpecs: execinfrapb.StreamIngestionPartitionSpecs{
				Specs: []*execinfrapb.StreamIngestionPartitionSpec{
					{
						SrcInstanceID:  1,
						DestInstanceID: 2,
						Spans: []roachpb.Span{
							{Key: roachpb.Key("a"), EndKey: roachpb.Key("c")},
						},
					},
				}},
			frontierEntries: execinfrapb.FrontierEntries{ResolvedSpans: []jobspb.ResolvedSpan{
				{
					Span:      roachpb.Span{Key: roachpb.Key("a"), EndKey: roachpb.Key("b")},
					Timestamp: hlc.Timestamp{WallTime: 1},
				},
{
					Span:      roachpb.Span{Key: roachpb.Key("b"), EndKey: roachpb.Key("c")},
					Timestamp: hlc.Timestamp{WallTime: 2},
				},
			}},
```

msbutler reviewed (commented):
Looks great! left mosly nits and clarification questions. I also noticed that you didn't test a round trip of writing/reading this new data to/from the job info table. Could you add a test for that, or is it coming in a future PR?

stevendanna reviewed (commented):
Just ack'ing that I've seen this PR. I think @msbutler and @dt are probably enough reviewers here. Thanks for working on this.

adityamaru commented on pkg/ccl/streamingccl/streamingest/stream_ingestion_dist.go:
```diff
@@ -190,6 +191,31 @@ func (p *replicationFlowPlanner) containsInitialStreamAddresses() bool {
+		if err != nil {
+			return err
+		}
+		return jobInfoStorage.Write(ctx, replicationPartitionInfoKey, specBytes)
```
this was a good callout. Unfortunately, the `Write` and `Read` methods in their current form were reserved for those files that we wanted to show up in the `Advanced Debugging` tab on the jobs details page. This is because they implicitly wrote an info row to a `~profiler/` prefixed info key. I have added a comment to this effect, but also pulled out the core of this method into its own utility so that other callsites that just want to write a file to job_info and read it back can use. Lets get https://github.com/cockroachdb/cockroach/pull/110035 merged before I rebase and fix this PR up?

adityamaru replied:
fixed

adityamaru commented on pkg/ccl/streamingccl/streamingest/stream_ingestion_frontier_processor.go:
changed to use `WriteChunkedExecutionDetailFile`

adityamaru commented:
finally got back to this, @msbutler to recap:
- i've fixed up the tests as per your comments
- i've deleted the timer based loop in the resumer, now the file is only generated on demand in the new `CollectProfile` method. We still have one time in how frequently we snapshot and dump the frontier but I don't think anything can be done about that right now. It defaults to 10mins.
- i've added an e2e test that compares the generated file against a baked in one.
- i've moved all direct calls to JobInfo to use the Read/Write helpers we checked in separately.

msbutler commented on pkg/ccl/streamingccl/streamingest/replication_execution_details.go:
```diff
@@ -0,0 +1,189 @@
+func generateSpanFrontierExecutionDetailFile(
+	ctx context.Context, execCfg *sql.ExecutorConfig, ingestionJobID jobspb.JobID, skipBehindBy bool,
+) error {
+	var sb bytes.Buffer
```
given the possibility of txn retries, i think `sb` and `w` should be created within the txn, right?

adityamaru replied:
you are very right, nice catch

msbutler commented on pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go:
```diff
@@ -536,8 +536,16 @@ func (s *streamIngestionResumer) CollectProfile(ctx context.Context, execCtx int
-	return bulkutil.FlushTracingAggregatorStats(ctx, s.job.ID(),
-		p.ExecCfg().InternalDB, aggStatsCopy)
+	if err := bulkutil.FlushTracingAggregatorStats(ctx, s.job.ID(),
+		p.ExecCfg().InternalDB, aggStatsCopy); err != nil {
```
should we instead log errors if any of these independent profiling tasks fail? I.e. if the TracingAggregatorStats call fails, it would still be nice to try get the FrontierExectionDetailFile

adityamaru replied:
good point, I still think I want the final error to bubble up to the client so that we know something is wrong but it makes sense to do everything we can before that. How do you like this?

msbutler commented on pkg/ccl/streamingccl/streamingest/stream_ingestion_dist.go:
```diff
@@ -56,6 +55,10 @@ var replanFrequency = settings.RegisterDurationSetting(
 
+// replicationPartitionInfoFilename is the filename at which the replication job
+// resumer writes its partition specs.
+const replicationPartitionInfoFilename = "~replication-partition-specs.binpb"
```
nit: is this used? if not, sad the linters did not catch this...

adityamaru replied:
it is! in `persistStreamIngestionPartitionSpecs` and the read path when generating the file. Its where we read/write the replication partition specs.

msbutler reviewed (commented):
Looks good! only blocking comment is the txn retry question.

adityamaru commented:
TFTR!

bor r=msbutler

adityamaru commented:
bors r=msbutler
